R/test-directory-c.R

```diff
@@ -0,0 +1,356 @@
+  queue <- start_test_processes(
+    dir = path,
+    paths = paths,
+    pkg_name = pkg_name,
+    testthat_dir = testthat_dir,
+    load_helpers = load_helpers,
+    pkgload = pkgload
+  )
+
+  # Run the event loop
+
+  # Need to remove the environment and source references, otherwise
+  # we need to copy a lot of data to the subprocess.
+
+  fun <- function(path) asNamespace("testthat")$subprocess_task(path)
+  environment(fun) <- .GlobalEnv
+  fun <- utils::removeSource(fun)
+  for (path in paths) {
+    queue$push(fun, list(path))
+  }
+
+  files <- list()
+
+  lister <- ListReporter$new()
+
+  replay <- function(filename) {
+    rtr <- get_reporter()
+    for (event in files[[filename]]) {
+      do.call(rtr[[event$cmd]], event$args)
+      do.call(lister[[event$cmd]], event$args)
+    }
+    files[[filename]] <- NULL
+  }
+
+  with_reporter(reporter,
+    while (!queue$is_idle()) {
+      msgs <- queue$poll(Inf)
+      for (x in msgs) {
+        if (x$code == 301) {
+          m <- x$message
+          # TODO: should we keep other messages in the subprocess? Probably.
+          if (!inherits(m, "testthat_message")) next
+          files[[m$filename]] <- append(files[[m$filename]], list(m))
+          if (m$cmd == "end_file") replay(m$filename)
+
+        } else if (x$code == 200) {
+          # File is done, nothing to do here, we'll get an end_file
+
+        } else if (x$code > 500) {
+          # a subprocess has crashed
+          # TODO: start another one. Here or in the task queue?
+          # TODO: print the partial result and report the crash
+        }
+      }
+    }
+  )
+
+  results <- lister$get_results()
+
+  if (stop_on_failure && !all_passed(results)) {
+    stop("Test failures", call. = FALSE)
+  }
+
+  if (stop_on_warning && any_warnings(results)) {
+    stop("Tests generated warnings", call. = FALSE)
+  }
+
+  invisible(results)
+}
+
+# This is the function that is called for each test file, in the subprocess
+
+subprocess_task <- function(path) {
+  on.exit(teardown_run(dirname(path)), add = TRUE)
+  env <- .GlobalEnv$.test_env
+  reporter <- find_reporter("subprocess")
+
+  with_reporter(reporter = reporter, code = {
+    reporter$start_file(basename(path))
+    source_file(
+      path, new.env(parent = env),
+      chdir = TRUE, wrap = TRUE
+    )
+    reporter$end_file()
+  })
+
+  NULL
+}
+
+find_pkg_name <- function(dir) {
+  desc_path <- file.path(dir, "DESCRIPTION")
+  if (!file.exists(desc_path)) {
+    stop("no DESCRIPTION file at path '", desc_path, "'", call. = FALSE)
+  }
+
+  desc <- read.dcf(desc_path, all = TRUE)
+  pkg <- desc$Package
+  if (!nzchar(pkg)) {
+    stop("no 'Package' field in DESCRIPTION file '", desc_path, "'", call. = FALSE)
+  }
+  pkg
+}
+
+start_test_processes <- function(dir, paths, pkg_name, testthat_dir,
+                                 load_helpers, pkgload) {
+
+  # TODO: detect number of CPUs
+  num_proc <- min(getOption("Ncpus", 4), length(paths))
```

Comment:
Yeah, this is temporary, see TODO list in the PR. 

The `Config/testthat/processes` is a bit weird, because the right number depends more on the machine than on the package. Well, unless you want to fix it to one, because you tests concurrently. I think such an option makes sense, but we would need to combine it with an option set by the user, I think.